### ISSUE TITLE
some metrics question

### ISSUE THREAD

**Issue by neverstoplearn** [mpapadomanolaki/UNetLSTM#2]
your paper report the best result is precision 63.59 recall 52.93 and f1 96.00 and oa 57.78, but your code just calculate cofuseion_matrix,i use the cofuseion_matrix.conf to calculate precision recall f1 oa, I can got precision: 0.9947 recall: 0.9610 f_score: 0.9775 oa: 0.9610,i am not sure if is right,can you tell me how you calculate precision recall f1 oa? thanks very much.

**Comment by mpapadomanolaki** [comment]
Hello,

I am calculating the metrics using the TP, TN, FP and FN pixels as explained in here https://towardsdatascience.com/beyond-accuracy-precision-and-recall-3da06bea9f6c . 
Precision, recall and F1 score can be calculated for each one of the two classes (change or no change). Your metric results seem to me that they concern the no change class. Make sure that you calculate the metrics for the change class.

Best,
Maria

**Comment by neverstoplearn** [comment]
I add
 `
def Metrics(data):
        tn = data[1][1]
        fn = data[0][1]
        fp = data[1][0]
        tp = data[0][0]

        precision =  tp / (tp + fp)

        recall =  tp / (tp + fn)

        pr = 2 *(tp / (tp + fp)) * (tp / (tp + fn))
        p_r = (tp / (tp + fp)) + (tp / (tp + fn))
        f_score = pr / p_r

        t_pn = tp + tn
        t_tpn = tp + tn + fp + fn
        oa =  t_pn / t_tpn
        return precision, recall, f_score, oa
`
      to calculate the precision recall f_score oa after
`for i, batch, in enumerate(tqdm(mydataset)):
        img_batch, lbl_batch = batch
        img_batch, lbl_batch = tools.to_cuda(img_batch.permute(1,0,2,3,4)), tools.to_cuda(lbl_batch)
        optimizer.zero_grad()
        output=model(img_batch.float())
        output_conf, target_conf = tools.conf_m(output, lbl_batch)
        confusion_matrix.add(output_conf, target_conf)
        loss=criterion(output, lbl_batch.long())
        train_losses.append(loss.item())
        loss.backward()
        optimizer.step()
        _, preds = output.data.max(1)
        if iter_ % 100 == 0:
         pred = np.argmax(output.data.cpu().numpy()[0], axis=0)
         gt = lbl_batch.data.cpu().numpy()[0]
         print('Train (epoch {}/{}) [{}/{} ({:.0f}%)]\tLoss: {:.6f}\tAccuracy: {}'.format(
                      epoch, epochs, i, len(mydataset),100.*i/len(mydataset), loss.item(), tools.accuracy(pred, gt)))
        iter_ += 1
        del(img_batch, lbl_batch, loss)
    #print(confusion_matrix.conf)
    #print(confusion_matrix)
    **precision, recall, f_score, oa = Metrics(confusion_matrix.conf)**`
does the confusion_matrix.conf means the unchange class ?

**Comment by mpapadomanolaki** [comment]
Now you are calculating the metrics for the no change class. To find the metrics for the change class you could rearrange the given matrix, or just calculate them like:

precision = data[1][1] / (data[1][1] + data[0][1])

recall = data[1][1] / (data[1][1] + data[1][0])

**Comment by neverstoplearn** [comment]
i almost know what you mean. thanks for your help.

**Comment by neverstoplearn** [comment]
I follow you advise, change the code:
`
def Metrics2(data):
        tp = data[1][1]
        fp = data[0][1]
        fn = data[1][0]
        tn = data[0][0]

        precision =  tp / (tp + fp)

        recall =  tp / (tp + fn)

        pr = 2 *(tp / (tp + fp)) * (tp / (tp + fn))
        p_r = (tp / (tp + fp)) + (tp / (tp + fn))
        f_score = pr / p_r

        t_pn = tp + tn
        t_tpn = tp + tn + fp + fn
        oa =  t_pn / t_tpn
        return precision, recall, f_score, oa`
but it also got   epoch: 60 precision: 0.787  recall: 0.938  f_score: 0.856 oa: 0.963 in val dataset
